Message versioning and ROS 2 message translation (#3465)
PX4/PX4-user_guide: en/ros2/px4_ros2_msg_translation_node.md

# PX4 ROS 2 Message Translation Node

<Badge type="tip" text="main (PX4 v1.16+)" /> <Badge type="warning" text="Experimental" />

> The message translation node allows to run applications that are compiled with one set of message versions against a PX4 with another set of message versions, without having to change either the application or the PX4 side.

Reviewer: ```suggestion
The message translation node allows applications that were compiled against different versions of the PX4 messages to interwork with newer versions of PX4, and vice versa, without having to change either the application or the PX4 side.
```

> Specifically for this to work, topic publication/subscription/service names contain a message version in the form of `<topic_name>_v<version>`.

Reviewer: I don't understand what this means. Specifically, at this point if I look at https://github.com/PX4/PX4-Autopilot/blob/main/msg/versioned/VehicleLandDetected.msg it does not have this format - I suspect this is all ROS2 stuff.

I don't think you need to include this line here. I would include information about how this works lower down, maybe in the implementation section. Not needed here.

> The translation node knows about all existing message versions, and dynamically monitors the publications, subscriptions and services, and then creates translations as needed.

Reviewer: ```suggestion
The translation node knows about all message versions previously defined by PX4, and dynamically monitors the publications, subscriptions and services, in order to create translations as needed.
```

> ## Installation and First Test

Reviewer: First test is implied IMO.

```suggestion
## Installation
```

> 

Reviewer: I think it is worth saying when you should do this. I presume for every workspace? 

So something like: "For every workspace you create, it is recommended to install and run the node:", followed by the steps.

Reviewer: Ah, so you only need the translation node running in one workspace - this observes the DDS data space. So you could run this node in the same workspace as an application or a different workspace. Right?

I see that in the User guide you have included the translation node in the same workspace as the example workspace you are building. I took that to mean "if you build a workspace and it needs translation, then the node must be running in that workspace" AND "if you have multiple workspaces that need translation, they would each need to run the note.

So what I would do is leave this page alone.

In the **user guide** I would remove the Optional steps you added into their own section "Starting the translation node". I'd put that after the example and have some very simple text and a link to the installation section below. The point being that you want to forward reference that you might need to use a translation node, but actually this is not something you need to do for this example, and not something you need to do every workspace.

The text might start something like ...

This example is built with PX4 and ROS2  versions that use the same message definitions. 
If you were to use incompatible [message versions] you would need to install and run a [translation node] as well, before running the example.

Does that make sense?

Reviewer: PS If this only works with ROS Jazzy, we should say that too, and it makes more sense to do that in a separate section.

Author: >  So you could run this node in the same workspace as an application or a different workspace. Right?

Correct

> Does that make sense?

It does, I've added such a section in the ROS 2 User Guide 👍🏻

> Create a ROS 2 workspace in which to build the message translation node and its dependencies.
> 
> ```bash
> mkdir -p /path/to/ros_ws/src
> ```
> 
> Run the following helper script, it will copy the message definitions and translation node to your ROS workspace directory.
> 
> ```bash
> cd /path/to/ros_ws
> /path/to/PX4-Autopilot/Tools/copy_to_ros_ws.sh .
> ```
> 
> Build and source the workspace.
> 
> ```bash
> colcon build
> source /path/to/ros_ws/install/setup.bash
> ```
> 
> Finally, run the translation node. You should see an output similar to:
> 
> ```bash
> ros2 run translation_node translation_node_bin
> [INFO] [1734525720.729530513] [translation_node]: Registered pub/sub topics and versions:
> [INFO] [1734525720.729594413] [translation_node]: Registered services and versions:
> ```

Reviewer: I like lists, in particular if you add the prefix I suggest above. So

````suggestion
1. Create a ROS 2 workspace in which to build the message translation node and its dependencies.

   ```sh
   mkdir -p /path/to/ros_ws/src
   ```

2. Run the following helper script to copy the message definitions and translation node into your ROS workspace directory.

   ```sh
   cd /path/to/ros_ws
   /path/to/PX4-Autopilot/Tools/copy_to_ros_ws.sh .
   ```

3. Build and source the workspace.

   ```sh
   colcon build
source /path/to/ros_ws/install/setup.bash
   ```
4. Finally, run the translation node:

   ```sh
   ros2 run translation_node translation_node_bin
   ```

    You should see an output similar to:

   ```sh
   [INFO] [1734525720.729530513] [translation_node]: Registered pub/sub topics and versions:
   [INFO] [1734525720.729594413] [translation_node]: Registered services and versions:
   ```
````

Author: Awesome thanks, I agree the list is much neater. Regarding the prefix what do you mean, are you referring to "For every workspace you create, it is recommended to install and run the node:"?

Reviewer: Yes!

## Updating a Message

When changing a message, a new version needs to be added.

> The steps include:

Reviewer: ```suggestion
The steps are:
```

> - Copy the versioned message file (`.msg`/`.srv`) to `px4_msgs_old/msg/` (or `px4_msgs_old/srv/`) and add the current version to the file name.

Reviewer: > Copy the versioned message file (`.msg`/`.srv`)`

- what is .srv here?
- what is `px4_msgs_old`? How do they get that?

  For example `msg/versioned/VehicleAttitude.msg` becomes `px4_msgs_old/msg/VehicleAttitudeV3.msg`.

>   Update the existing translations that use the current topic version to the now old version.

Reviewer: > Update the existing translations that use the current topic version to the now old version.

1. What's a translation? Needs a definition.
2. Am I replacing or adding a new one of these "whatever they are". I guess I'm asking if I make 3 versions of my message do I have three files? Or one that is updated three times?

Author: Good points, I've elaborated on this.

> I guess I'm asking if I make 3 versions of my message do I have three files?

Exactly, 3 new versions of a message means adding three new translation files: each translation file bridging one version to the next

>   For example `px4_msgs::msg::VehicleAttitude` becomes `px4_msgs_old::msg::VehicleAttitudeV3`.

Reviewer: This example doesn't help answer my questions above  - in what file would it be?

Author: The existing translations would be under `msg/translation_node/translations/*.h`. I've reworked the section to make it clearer

> - Increment `MESSAGE_VERSION` and update the message fields as desired.

Reviewer: ```suggestion
- Increment `MESSAGE_VERSION` in the current message definition (the one in `msg/versioned/`).
  Make the other changes to message fields that prompted the version change.
```

Reviewer: Note ^^^ Just trying to be precise about which file/message is being updated.

  - [`translations/example_translation_multi_v2.h`](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/example_translation_multi_v2.h)
  - [`translations/example_translation_service_v1.h`](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/example_translation_service_v1.h)
- Include the added header in [`translations/all_translations.h`](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/all_translations.h).

For the second last step and for topics, there are two options:

1. Direct translations: these translate a single topic between two different versions.
  This is the simpler case and should be preferred if possible.
2. Generic case: this allows a translation between N input topics and M output topics.
  This can be used for merging or splitting a message.
  Or for example when moving a field from one message to another, a single translation should be added with the two older message versions as input and the two newer versions as output.

>   This way there is no information lost when translating forward or backward.

Reviewer: I haven't looked at the example, but it seems to me that this might require some worked examples of each type. If the headers are specific to each type, perhaps move them under the following sections?

Author: You're right, I've gone and added a worked example also to make the steps clearer

::: info
If a nested message definition changes, all messages including that message also require a version update.

> This is primarily important for services.

Reviewer: I don't know anything about nested message definitions. What are they? 

Either way, perhaps a warning or tip?

Author: Nested messages are message definitions that include fields referencing other message types instead of using only the classic data types. For instance [PositionSetpointTriplet](https://github.com/PX4/PX4-Autopilot/blob/main/msg/PositionSetpointTriplet.msg).

I've added the example and made the paragraph into a warning

Reviewer: Ah, probably we need to update https://docs.px4.io/main/en/middleware/uorb.html#adding-a-new-topic with a new heading on nesting - this mentions nesting, but has no example and is easy to miss. That would be separate job, but if you're able to to it would be appreciated.

Author: I missed this comment until now, I've opened a PR for it:
https://github.com/PX4/PX4-user_guide/pull/3562

:::

> 

Reviewer: What do I do if I have custom messages I want to version?

Author: That would not be a problem, those custom messages would just need the `MESSAGE_VERSION` field to be added and the file moved to the corresponding `versioned/` folder. Updating those message would follow the same steps as updating vanilla px4 messages

## Usage in ROS

The message version can be added generically to a topic like this:

```c++
topic_name + "_v" + std::to_string(T::MESSAGE_VERSION)
```

Where `T` is the message type, e.g. `px4_msgs::msg::VehicleAttitude`.

The DDS client in PX4 automatically adds the version suffix if a message contains the field `uint32 MESSAGE_VERSION = x`.

> ::: info Version 0 of a topic means that no `_v<version>` suffix should be added.
> :::

Reviewer: ```suggestion
::: info
Version 0 of a topic means that no `_v<version>` suffix should be added.
:::
```

- Include the added header in [`translations/all_translations.h`](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/all_translations.h).

For the second last step and for topics, there are two options:

1. Direct translations: these translate a single topic between two different versions.
  This is the simpler case and should be preferred if possible.
2. Generic case: this allows a translation between N input topics and M output topics.
  This can be used for merging or splitting a message.
  Or for example when moving a field from one message to another, a single translation should be added with the two older message versions as input and the two newer versions as output.
  This way there is no information lost when translating forward or backward.

::: info
If a nested message definition changes, all messages including that message also require a version update.
This is primarily important for services.
:::

## Usage in ROS

The message version can be added generically to a topic like this:

```c++

> topic_name + "_v" + std::to_string(T::MESSAGE_VERSION)

Reviewer: So this is something you have to do for all topics?

I think I might make this a bit more "real" by having a several line real-world example afterwards.

Version 0 of a topic means that no `_v<version>` suffix should be added.
:::

## Implementation Details

The translation node dynamically monitors the topics and services.
It then instantiates the countersides of the publications and subscribers as required.
For example if there is an external publisher for version 1 of a topic and subscriber for version 2.

Internally, it maintains a graph of all known topic and version tuples (which are the graph nodes).
The graph is connected by the message translations.
As arbitrary message translations can be registered, the graph can have cycles and multiple paths from one node to another.
Therefore on a topic update, the graph is traversed using a shortest path algorithm.
When moving from one node to the next, the message translation method is called with the current topic data.
If a node contains an instantiated publisher (because it previously detected an external subscriber), the data is published.
Thus, multiple subscribers of any version of the topic can be updated with the correct version of the data.

For translations with multiple input topics, the translation continues once all input messages are available.

## Limitations

> - The current implementation depends on a service API that is not yet available in ROS Humble, and therefore does not support services when built for ROS Humble.

Reviewer: So what versions will it work for?

Reviewer 2: Jazzy or later

Reviewer: @GuillaumeLaine So ROS Jazzy is a dependency - if this is never going to be in ROS Humble we should maybe say

```suggestion
- The current implementation depends on a service API that is present from ROS Jazzy (ROS applications build for Humble and earlier cannot be used with the node).
```

I added a version for ROS Jazzy above on this doc. We should also add it in the guide, as noted in https://github.com/PX4/PX4-user_guide/pull/3465/files#r1924601823

Author: @hamishwillee I think there is some confusion with this bullet-point, let me try to clarify:

- On ROS Humble and older, the translation node will work but only for the translation of topic messages, not services messages. Service message translation requires a service API that is present starting from ROS Jazzy
- On ROS Jazzy and later, the translation will work to its full extent, both topic and service translation. 

Currently translating topic messages is the main use-case for the translation node, so in that sense ROS Jazzy isn't a mandatory dependency. But clearly this point wasn't clear, I'll update the docs now

Reviewer: Ah, I didn't understand - that's pretty much because to me service messages are still a magical nothing that is undefined.

The fix here would be:

 ```suggestion
- Translation of service messages does not work on ROS Humble, but does on ROS Jazzy.
  This is because the current implementation depends on a service API that is not yet available in ROS Humble.
  Translation of topic messages is fully supported.
```

Then undo the badging I added for ROS Jazzy.

This is better because it starts by saying exactly what does not work, rather than the missing dependency.

Make sense?

> - Services only support a linear history, i.e. no message splitting or merging

Reviewer: Do you mean services as in something different than message/topics? Perhaps this is what you mean by .svr above? Is there information on what those mean anywhere?

Either way, so are you saying that services require a linear history, but you can split or merge messages as you like?

Author: Correct, ROS distinguishes between services and topics like so: https://docs.ros.org/en/foxy/How-To-Guides/Topics-Services-Actions.html. In PX4, we have the [srv/ services](https://github.com/PX4/PX4-Autopilot/tree/message_versioning_and_translation/srv) folder vs the [msg/ topics](https://github.com/PX4/PX4-Autopilot/tree/message_versioning_and_translation/msg) folder.

Messages define the data format used to communicate over _either_ topics or via services. The confusing part is that `.srv` files define service messages, and `.msg` files define topic messages (the file extension introduces confusion and make them seem more generic)

And finally yes, currently services require a linear history to translate while topics (`.msg`) can be split and merged

Thanks for the comment, I realized I was also confused by this. I'll do my best to make this clear in the docs

> - Having both publishers and subscribers for two different versions of a topic is currently not handled by the translation node and would trigger infinite circular publications.

Reviewer: So a ROS application must compile against just one version of a topic? What if there are two ROS applications running on the network compiled against different versions - is this the problem case?

Reviewer 2: > So a ROS application must compile against just one version of a topic?

Yes

> What if there are two ROS applications running on the network compiled against different versions - is this the problem case?

No, only a specific subset, like:
- app 1: pub topic_v1, sub topic_v1
- app 2: pub topic_v2, sub topic_v2

In practice there's rarely (if any) setup that will run into this.

Reviewer: > Having both publishers and subscribers for two different versions of a topic is currently not handled by the translation node and would trigger infinite circular publications.

OK, I think I get it. This fails because you have "both publishers and subscribers for two different versions of a topic"
```
app 1: pub topic_v1, sub topic_v1
app 2: pub topic_v2, sub topic_v2
```
But this would not fail, because you have publishers of a different topic, but all subscribers subscribe to the same topic.
```
app 1: pub topic_v1,  sub topic_v2
app 2: pub topic_v2, sub topic_v2
```

Reviewer 2: Exactly

Reviewer: Given I had so much trouble with this, perhaps insert the example? YOur call.

````suggestion
- Having both publishers and subscribers for two different versions of a topic is currently not handled by the translation node and would trigger infinite circular publications:

  ```
  app 1: pub topic_v1, sub topic_v1
  app 2: pub topic_v2, sub topic_v2
  ```

  Note that the far more common case, where you have multiple apps subscribing to the same topic and different publishers (or vice versa), are fine:
  ```

  app 1: pub topic_v1, sub topic_v2
  app 2: pub topic_v2, sub topic_v2
  ```

````

Author: I've added the example, it certainly helps clarify.

I've changed your suggestion a bit; the reason why the circular configuration is unlikely is because ROS topics from the FMU (carrying the messages that might be translated) are intended to be directional (i.e. /fmu/out vs /fmu/in). So the common case would be even simpler, more like:
```
  app 1: pub topic1_v1, sub topic2_v3
  app 2: pub topic1_v2, sub topic2_v5
```
i.e. apps don't sub to topic1, apps don't pub to topic2

And just to avoid any misconception, I think your 2nd example isn't possible:
  ```
  app 1: pub topic_v1, sub topic_v2
  app 2: pub topic_v2, sub topic_v2
  ```
Namely an app only has a single definition/version for a message, so `app1` could not subscribe to the same topic as `v1` and publish as `v2`

Reviewer: Thanks - your explanation is much better than mine (obviously!).

# PX4 ROS 2 Message Translation Node

<Badge type="tip" text="main (PX4 v1.16+)" /> <Badge type="warning" text="Experimental" />

The message translation node allows ROS 2 applications that were compiled against different versions of the PX4 messages to interwork with newer versions of PX4, and vice versa, without having to change either the application or the PX4 side.

## Overview

> The translation of messages from one definition version to another is possbile thanks to the introduction of [message versioning](../middleware/uorb.md#message-versioning).

Reviewer 3: ```suggestion
The translation of messages from one definition version to another is possible thanks to the introduction of [message versioning](../middleware/uorb.md#message-versioning).
```

> The translation node has access to all message versions previously defined by PX4. It dynamically observes the DDS data space, monitoring the publications, subscriptions and services originating from either PX4 via the [uXRCE-DDS Bridge](../middleware/uxrce_dds.md), or ROS2 applications. When necessary, it converts messages to the current versions expected by both applications and PX4, ensuring compatibility.

Reviewer 3: ```suggestion
The translation node has access to all message versions previously defined by PX4. It dynamically observes the DDS data space, monitoring the publications, subscriptions and services originating from either PX4 via the [uXRCE-DDS Bridge](../middleware/uxrce_dds.md), or ROS 2 applications. When necessary, it converts messages to the current versions expected by both applications and PX4, ensuring compatibility.
```

> ![Overview ROS 2 Message Translation Node](../../assets/middleware/ros2/px4_ros2_interface_lib/translation_node.svg)

Reviewer: I understand this diagram, but it took me a little while to understand what it meant - specifically what versions are supported. Do you think this makes it easier?

![image](https://github.com/user-attachments/assets/770cea06-5387-4471-824f-84c62bb7310a)

Reviewer: Yes, that's the only change. 

Initially I was thinking of marking up the nodes with the versions of messages that they sub/pub because I wanted to be able to show what the translation node does and why. With colour coding I think we can avoid this.

  public:
    MinimalPubSub() : Node("minimal_pub_sub") {
      // Use template function to define the correct topics automatically
      const std::string sub_topic = "/fmu/out/vehicle_attitude" + getMessageNameVersion<px4_msgs::msg::VehicleAttitude>();
      const std::string pub_topic = "/fmu/in/vehicle_command" + getMessageNameVersion<px4_msgs::msg::VehicleCommand>();

      _subscription = this->create_subscription<px4_msgs::msg::VehicleAttitude>(
          sub_topic, 10,
          std::bind(&MinimalPubSub::attitude_callback, this, std::placeholders::_1));
      _publisher = this->create_publisher<px4_msgs::msg::VehicleCommand>(pub_topic, 10);
    }

  private:
    void attitude_callback(const px4_msgs::msg::VehicleAttitude::SharedPtr msg) {
      RCLCPP_INFO(this->get_logger(), "Received attitude message.");
    }

    rclcpp::Publisher<px4_msgs::msg::VehicleCommand>::SharedPtr _publisher;
    rclcpp::Subscription<px4_msgs::msg::VehicleAttitude>::SharedPtr _subscription;
};
```

:::

::: tab Python

```python
import rclpy
from rclpy.node import Node
from px4_msgs.msg import VehicleCommand, VehicleAttitude

# Helper function to get the message version suffix

> # The correct message version is directly inferred from the message defintion

Reviewer 3: ```suggestion
# The correct message version is directly inferred from the message definition
```

            sub_topic,
            self.attitude_callback,
            10
        )

        self._publisher = self.create_publisher(
            VehicleCommand,
            pub_topic,
            10
        )

    def attitude_callback(self, msg):
        self.get_logger().info("Received attitude message.")
```

:::

::::

On the PX4 side, the DDS client automatically adds the version suffix if a message definition contains the field `uint32 MESSAGE_VERSION = x`.

::: info
Version 0 of a topic means that no `_v<version>` suffix should be added.
:::

## Development

### Definitions

A __message__ defines the data format used for communication, whether over a topic or a service.
Therefore a message can be either a _topic_ message defined by a `.msg` file, or a _service_ message defined by a `.srv` file.

A __versioned message__ is a message for which changes are tracked and each change results in a version bump, with the previous state of the definition being stored in history.
The latest version of every message is stored in `msg/versioned/` for topics (or `srv/versioned` for services), and all older versions are stored in `msg/px4_msgs_old/msg/` (or `msg/px4_msgs_old/srv/`). 

> A __version translation__ defines a bidrectional mapping of the contents of one or more message definition across different versions.

Reviewer 3: ```suggestion
A __version translation__ defines a bidirectional mapping of the contents of one or more message definition across different versions.
```

└── srv/
  ├── *.srv              # Non-versioned service message files
  └── versioned/         # Latest versioned service message files
```

This structure introduces new directories: `versioned/`, `px4_msgs_old/`, and `translation_node/`.

#### Directories `msg/versioned/` and `srv/versioned/`

- Contain the current latest version of each message.
- Files in these directories must include a `MESSAGE_VERSION` field to indicate that they are versioned.
- File names follow the conventional naming scheme (without a version suffix).

Example directory structure:

```
PX4-Autopilot
├── ...
├── msg/
  └── versioned/
    ├── VehicleAttitude.msg        # e.g. MESSAGE_VERSION = 3
    └── VehicleGlobalPosition.msg  # e.g. MESSAGE_VERSION = 2
└── srv/
  └── versioned/
    └── VehicleCommand.srv         # e.g. MESSAGE_VERSION = 2
```

#### Directory `px4_msgs_old/`

- Archives the history of all versioned messages, including both topic and service messages (resp. under `msg/` and `srv/` subdirectories).
- Each file includes a `MESSAGE_VERSION` field.
- File names reflect the message's version with a suffix (e.g., `V1`, `V2`).

> Example directory structure (coherant with example above):

Reviewer 3: ```suggestion
Example directory structure (coherent with example above):
```

```
  ...
  msg/
  └── px4_msgs_old/
    ├── msg/
      ├── VehicleAttitudeV1.msg
      ├── VehicleAttitudeV2.msg
      └── VehicleGlobalPositionV1.msg
    └── srv/
      └── VehicleCommandV1.srv
```

#### Directory `translation_node/`

- Contains headers for translating between all different versions of messages.
- Each translation (direct or generic) is a single `.h` header file.
- The header `all_translation.h` acts as the main header, and includes all subsequent translation headers.

> Example directory structure (coherant with example above):

Reviewer 3: ```suggestion
Example directory structure (coherent with example above):
```

It then instantiates the counterside of the publications and subscribers as required.
For example if there is an external publisher for version 1 of a topic and subscriber for version 2.

Internally, it maintains a graph of all known topic and version tuples (which are the graph nodes).
The graph is connected by the message translations.
As arbitrary message translations can be registered, the graph can have cycles and multiple paths from one node to another.
Therefore on a topic update, the graph is traversed using a shortest path algorithm.
When moving from one node to the next, the message translation method is called with the current topic data.
If a node contains an instantiated publisher (because it previously detected an external subscriber), the data is published.
Thus, multiple subscribers of any version of the topic can be updated with the correct version of the data.

For translations with multiple input topics, the translation continues once all input messages are available.

## Limitations

- The current implementation depends on a service API that is not yet available in ROS Humble, and therefore does not support services when built for ROS Humble.

> - Services only support a linear history, i.e. no message splitting or merging.

Reviewer: I think this means 
```suggestion
- Service messages only support a linear history, i.e. no message splitting or merging.
```

It matters because _services_ is a very overloaded term. We should at least make it specific so anyone who doesn't understand what a service message is (i.e. me) knows they can ignore this limitation.

Example directory structure (coherent with example above):

```
  ...
  msg/
  └── translation_node/
    └── translations/
      ├── all_translations.h                        # Main header
      ├── translation_vehicle_attitude_v1.h         # Direct translation v0 <-> v1
      ├── translation_vehicle_attitude_v2.h         # Direct translation v1 <-> v2
      ├── translation_vehicle_attitude_v3.h         # Direct translation v2 <-> latest (v3)
      ├── translation_vehicle_global_position_v1.h  # Direct translation v0 <-> v1
      ├── translation_vehicle_global_position_v2.h  # Direct translation v1 <-> latest (v2)
      ├── translation_vehicle_command_v1.h          # Direct translation v0 <-> v1
      └── translation_vehicle_command_v2.h          # Direct translation v1 <-> latest (v2)
```

### Updating a Versioned Message

This section provides a step-by-step walkthrough and a basic working example of what the process of changing a versioned message looks like.

Before making changes to a versioned message (in `msg/versioned/` or `srv/versioned/`), we first create an archived copy of it, and update existing translation code to point to the archived version definition.

> We then update versioned message definition to increment the version number, add any new fields or other changes, and then add a new translation file to convert between it and the previous (now archived) version.

Reviewer: ```suggestion
We then update the versioned message definition to increment the version number, add any new fields or other changes, and then add a new translation file to convert between it and the previous (now archived) version.
```

> We then update the versioned message definition to increment the version number, add any new fields or other changes, and then add a new translation file to convert between it and the previous (now archived) version.

Author: You're right I'm deleting it

   - [Generic Message Translation Template](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/example_translation_multi_v2.h)
   - [Direct Service Translation Template](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/example_translation_service_v1.h)

1. **Include New Headers in `all_translations.h`**

   <!-- TODO: update GitHub URL -->

   Add all newly created headers to [`translations/all_translations.h`](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/all_translations.h) so that the translation node can find them.

   For example, append the following line to `all_translation.h`:

   ```c++
   #include "translation_vehicle_attitude_v4.h"
   ```

Note that in the example above and in most cases, step 4 only requires the developer to create a direct translation for the definition change.
This is because the changes only involved a single message.

> In more complex cases of splitting, merging and/or moving definitions then a generic translation must be created.

Reviewer: It would be good to have an example of a generic translation and a service translation. This could be minimal, or a pointer to an example in the codebase once one exists.

Author: A few lines above we provide links to "version translation templates" which are example translation header files for each scenario: direct topic translation, generic topic translation, and direct service translation. 

Did you have something more elaborate in mind?

Reviewer: A little more ...

I was thinking perhaps show code here for the generic case. But as you imply, actually the template shows exactly this case you have outlined in the example below except the actual moving of the field.

> For example when moving a field from one message to another, a single generic translation should be added with the two older message versions as input, and the two newer versions as output.
> This ensures there is no information lost when translating forward or backward.

The problem is that this wasn't obvious to me, perhaps because I am not a developer. The only comment was  `// Translate ExampleTopic and OtherTopic v1 <--> v2` but that was a bit terse for me.

So perhaps after the "For example" add a comment like:

> This is exactly the approach shown by the [Generic Topic Message Translation Template](https://github.com/PX4/PX4-Autopilot/blob/message_versioning_and_translation/msg/translation_node/translations/example_translation_multi_v2.h), omitting only the code for actually modifying fields in the `fromOlder()` and `toOlder()` methods.

Author: Good point, I've applied this change to connect the paragraph to the code example

  This is because the current implementation depends on a service API that is not yet available in ROS Humble.
  Translation of topic messages is fully supported.
- Services messages only support a linear history, i.e. no message splitting or merging.

> - Having both publishers and subscribers for two different versions of the same topic is currently not handled by the translation node and would trigger infinite circular publications.
>   This refers to the following problematic configuration:
>   ```
>   app 1: pub topic_v1, sub topic_v1
>   app 2: pub topic_v2, sub topic_v2
>   ```
>   In practice this configuration is unlikely to occur because ROS topics shared with the FMU are intended to be directional (e.g. `/fmu/out/vehicle_status` or `/fmu/in/trajectory_setpoint`), therefore apps typically do not publish and subscribe simultaneously to the same topic.
>   The translation node could be extended to handle this corner-case if required.

Reviewer: ````suggestion
- Having both publishers and subscribers for two different versions of the same topic is currently not handled by the translation node and would trigger infinite circular publications.
  This refers to the following problematic configuration:
  
  ```
  app 1: pub topic_v1, sub topic_v1
  app 2: pub topic_v2, sub topic_v2
  ```

  In practice this configuration is unlikely to occur because ROS topics shared with the FMU are intended to be directional (e.g. `/fmu/out/vehicle_status` or `/fmu/in/trajectory_setpoint`), therefore apps typically do not publish and subscribe simultaneously to the same topic.
  The translation node could be extended to handle this corner-case if required.
````
